### PR TITLE
ARROW-312: Read and write Arrow IPC file format from Python

### PR DESCRIPTION
**Pull request by wesm**
This also adds some IO scaffolding for interacting with `arrow::Buffer` objects from Python and assorted additions to help with testing.

### CODE REVIEW

File: python/pyarrow/array.pyx [hunk 1]
```diff
@@ -37,8 +37,7 @@ import pyarrow.schema as schema
 
 
 def total_allocated_bytes():
-    cdef MemoryPool* pool = pyarrow.GetMemoryPool()
-    return pool.bytes_allocated()
+    cdef MemoryPool* pool = pyarrow.get_memory_pool()
```
Comment (xhochy) [inline python/pyarrow/array.pyx]: Accidential deletion?

**Review by xhochy — CHANGES_REQUESTED** [review]

File: python/pyarrow/array.pyx (hunk 1 above)
Comment (wesm) [inline python/pyarrow/array.pyx]: yep, fixed

**Comment by xhochy** [comment]
Locally CMake reports the following warning, this might be related to the OSX build failure in Travis:

```-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   arrow_ipc```

**Review by xhochy — APPROVED** [review]
Apart from the OSX failure: LGTM

**Comment by wesm** [comment]
Got it, that makes sense. Let me try to fix

**Comment by wesm** [comment]
There's a lot of duplicated boilerplate for configuring shared libs -- we should create some cmake functions to make this easier / cleaner -- see ARROW-330

**Comment by wesm** [comment]
+1

**Pull request closed without merging**